src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -51,7 +51,25 @@ isLeft (Right _) = False
 desugarCasesModule :: (Functor m, Applicative m, MonadSupply m, MonadError MultipleErrors m) => [Module] -> m [Module]
 desugarCasesModule ms = forM ms $ \(Module ss coms name ds exps) ->
   rethrow (addHint (ErrorInModule name)) $
-    Module ss coms name <$> (desugarCases <=< desugarAbs $ ds) <*> pure exps
+    Module ss coms name <$> (desugarCases <=< desugarAbs <=< validateCases $ ds) <*> pure exps
+
+-- |
+-- Validates that case head and binder lengths match.
+--
+validateCases :: (Functor m, Applicative m, MonadSupply m, MonadError MultipleErrors m) => [Declaration] -> m [Declaration]
```

Comment:
`base-compat` should make it possible to remove some of these constraints, I think.